src/core/extra.ml

```diff
@@ -100,6 +100,30 @@ module List =
       | e::es -> let fn e = Format.fprintf oc "%s%a" sep pp_elt e in
                  pp_elt oc e; iter fn es
 
+    (** [combine3 [a1;...; an] [b1;...; bn] [c1;...;cn]] returns
+        [(a1,b1,c1);...;(an,bn,cn)].
+        @raise Invalid_argument if the three lists are determined to have
+        different lengths. *)
+    let rec combine3 : 'a list -> 'b list -> 'c list -> ('a*'b*'c) list =
+      fun l1 l2 l3 ->
+      match l1, l2, l3 with
+      | [], [], [] -> []
+      | t1::q1, t2::q2, t3::q3 -> (t1,t2,t3)::(combine3 q1 q2 q3)
+      | _ ->
+          raise (Invalid_argument "Argument lists must have the same length")
+
+    (** [combine_rev [a1;...;an] [b1;...;bn]] returns [(an,bn);...;(a1,b1)].
+        @raise Invalid_argument if the two lists are determined to have
+        different lengths. *)
+    let combine_rev : 'a list -> 'b list -> ('a * 'b) list = fun l1 l2 ->
+      let rec aux l1 l2 acc = match l1, l2 with
+        | [], [] -> acc
+        | t1::q1, t2::q2 -> aux q1 q2 ((t1,t2)::acc)
+        | _ ->
+           raise (Invalid_argument "Argument lists must have the same length")
+      in
+      aux l1 l2 []
```

Comment:
This is List.fold_left2 (fun acc x y -> (x,y)::acc).

Reply:
Yes, you right.